common/tools/armageddon/broadcast_client.go

```diff
@@ -0,0 +1,226 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package armageddon
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+	"time"
+
+	"github.com/hyperledger/fabric-lib-go/common/flogging"
+
+	"github.com/hyperledger/fabric-protos-go-apiv2/common"
+	ab "github.com/hyperledger/fabric-protos-go-apiv2/orderer"
+	"github.com/hyperledger/fabric-x-orderer/node/comm"
+	"google.golang.org/grpc"
+)
+
+type StreamInfo struct {
+	conn                  *grpc.ClientConn
+	stream                ab.AtomicBroadcast_BroadcastClient
+	ctx                   context.Context
+	cancel                context.CancelFunc
+	isBroken              bool
+	isAlreadyReconnecting bool
+	maxRetryDelay         time.Duration
+	endpoint              string
+	lock                  sync.Mutex
+	logger                *flogging.FabricLogger
+}
+
+func (streamInfo *StreamInfo) IsBroken() bool {
+	streamInfo.lock.Lock()
+	defer streamInfo.lock.Unlock()
+	return streamInfo.isBroken
+}
+
+func (streamInfo *StreamInfo) SetIsBroken(value bool) {
+	streamInfo.lock.Lock()
+	defer streamInfo.lock.Unlock()
+	streamInfo.isBroken = value
+}
+
+func (streamInfo *StreamInfo) CheckIfReconnectionIsNeeded() bool {
+	streamInfo.lock.Lock()
+	defer streamInfo.lock.Unlock()
+	if streamInfo.isAlreadyReconnecting {
+		return false // reconnection is not needed again
+	} else {
+		streamInfo.isAlreadyReconnecting = true // mark that reconnection is needed
+		return true
+	}
+}
+
+func (streamInfo *StreamInfo) SetNewConnAndStream(newConnection *grpc.ClientConn, newStream ab.AtomicBroadcast_BroadcastClient) error {
+	streamInfo.lock.Lock()
+	defer streamInfo.lock.Unlock()
+	// close old connection
+	err := streamInfo.conn.Close()
```

Comment:
maybe close this before? when it is initially broken or before trying to reconnect?